### PR TITLE
C func call

### PR DESCRIPTION
**Pull request by sebasguts**

### CODE REVIEW

File: JuliaInterface/src/JuliaInterface.c [hunk 1]
```diff
@@ -207,6 +207,107 @@ Obj NewJuliaFunc(jl_function_t* function, int autoConvert)
...
 }
 
+/*
+ * C function pointer calls
+ */
+
+typedef Obj (*func_ptr_0_args)();
+Obj DoCallJuliaCFunc0Arg( Obj func )
+{
+    func_ptr_0_args function = (func_ptr_0_args)jl_unbox_voidpointer((jl_value_t*)FEXS_FUNC(func));
```
Comment (fingolfin) [inline JuliaInterface/src/JuliaInterface.c]: I would put `jl_unbox_voidpointer((jl_value_t*)FEXS_FUNC(func))` into a helper function. That'll make the code more readable, and also make it trivial to change how the func pointer is stored.

File: JuliaInterface/src/JuliaInterface.c [hunk 2]
```diff
@@ -207,6 +207,107 @@ Obj NewJuliaFunc(jl_function_t* function, int autoConvert)
...
+    return function(arg1,arg2,arg3,arg4,arg5,arg6);
+}
+
+Obj NewJuliaCFunc(void* function, int nr_args, char* args)
+{
+
+    Obj func = NewFunctionC("", nr_args, args, 0);
+
+    switch(nr_args){
+        case 0:
```
Comment (fingolfin) [inline JuliaInterface/src/JuliaInterface.c]: Off-topic, but: it'd be great if there was a `.clang-format` file in this repo. Then the code in this file could be formatted uniformly. I normally try to adapt to whatever code formatting conventions are used by a project, but here, I really can't tell what those are, resp. find them so different from any I used before that I have a hard time following them. In either case, a `.clang-format` would simplify writing patches for me a lot.

File: JuliaInterface/src/JuliaInterface.c [hunk 3]
```diff
@@ -207,6 +207,107 @@ Obj NewJuliaFunc(jl_function_t* function, int autoConvert)
...
+            SET_HDLR_FUNC(func, 3, DoCallJuliaCFunc3Arg);
+            break;
+        case 4:
+            SET_HDLR_FUNC(func, 4, DoCallJuliaCFunc4Arg);
+            break;
+        case 5:
+            SET_HDLR_FUNC(func, 5, DoCallJuliaCFunc5Arg);
+            break;
+        case 6:
+            SET_HDLR_FUNC(func, 6, DoCallJuliaCFunc6Arg);
```
Comment (fingolfin) [inline JuliaInterface/src/JuliaInterface.c]: While you can do that, the more "natural" way would be to pass the handler in the fourth argument to `NewFunctionC`. But I guess it doesn't really matter in the end.

File: JuliaInterface/src/JuliaInterface.c [hunk 4]
```diff
@@ -207,6 +207,107 @@ Obj NewJuliaFunc(jl_function_t* function, int autoConvert)
...
+            break;
+        default:
+            ErrorQuit("Only 0-6 arguments are supported", 0, 0);
```
Comment (fingolfin) [inline JuliaInterface/src/JuliaInterface.c]: Perhaps move this error to *before* calling `NewFunctionC`? Of course it shouldn't *really* matter thanks to garbage collection, but it just feels "more natural" to me to just always perform argument validation before using those arguments.

File: JuliaInterface/gap/BindCFunction.gi [hunk 5]
```diff
@@ -27,15 +27,16 @@ InstallGlobalFunction( JuliaBindCFunction,
         return;
     fi;
 
-    cfunction_call_string := Concatenation( "@cfunction(GAP.prepare_func_for_gap(", julia_name,")" );
```
Comment (fingolfin) [inline JuliaInterface/gap/BindCFunction.gi]: Also, is `prepare_func_for_gap` still used anywhere else? If not, remove it?

File: JuliaInterface/gap/BindCFunction.gi [hunk 6]
```diff
@@ -27,15 +27,16 @@ InstallGlobalFunction( JuliaBindCFunction,
...
     od;
-    Remove( cfunction_call_string );
-    cfunction_call_string := Concatenation( cfunction_call_string, "))" );
-
-    return _JuliaBindCFunction( cfunction_call_string, nr_args, arg_names );
+    cfunction_call_string := Concatenation( cfunction_call_string, "));" );
+    cfunction_call_string := Concatenation( cfunction_call_string, "\nreinterpret(Ptr{Cvoid},temp)");
+    function_ptr := JuliaEvalString( cfunction_call_string );
+    arg_string := JoinStringsWithSeparator( arg_names, "," );
+    return _NewJuliaCFunc( function_ptr, nr_args, arg_string );
```
Comment (fingolfin) [inline JuliaInterface/gap/BindCFunction.gi]: This code really could use some comments!

File: JuliaInterface/example/function_perform.gi [hunk 7]
```diff
@@ -0,0 +1,48 @@
+LoadPackage( "JuliaInterface" );
+
+dirs:= DirectoriesPackageLibrary( "JuliaInterface", "example" );
+
+JuliaIncludeFile( Filename( dirs, "function_perform.jl" ) );
+
+if not IsBound(_JuliaFunctionByModuleNoConv) then
+  _JuliaFunctionByModuleNoConv:= _JuliaFunctionByModule;
+  _JuliaFunctionNoConv:= _JuliaFunction;
```
Comment (fingolfin) [inline JuliaInterface/example/function_perform.gi]: This `if` block is not really needed.

And beyond that, as I mentioned elsewhere, I'd really merge `_JuliaFunctionByModuleNoConv`, `_JuliaFunctionByModule`, `_JuliaFunctionNoConv`, `_JuliaFunction` into a single function.

File: JuliaInterface/gap/BindCFunction.gi [hunk 8]
```diff
@@ -10,7 +10,7 @@
 
 InstallGlobalFunction( JuliaBindCFunction,
   function( julia_name, nr_args, arg_names )
-    local cfunction_call_string, i, cfunc;
+    local cfunction_call_string, i, cfunc, function_ptr, arg_string;
```
Comment (fingolfin) [inline JuliaInterface/gap/BindCFunction.gi]: Why does one need to specify `nr_args`? Isn't that always equal to `Length(arg_names)`?

File: JuliaInterface/example/function_perform.gi [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+blafuncC := JuliaBindCFunction( "blafunc", 2, ["a", "b"] );
+
+GASMAN("collect"); ListX([1..10^5], [1..10], {i,j} -> i);; time;
+GASMAN("collect"); ListX([1..10^5], [1..10], ReturnFirst);; time;
+GASMAN("collect"); ListX([1..10^5], [1..10], myfunc);; time;
+GASMAN("collect"); ListX([1..10^5], [1..10], blafunc);; time;
+# GASMAN("collect"); ListX([1..10^5], [1..10], myfuncC);; time;
+GASMAN("collect"); ListX([1..10^5], [1..10], blafuncC);; time;
+GASMAN("collect"); ListX([1..10^5], [1..10], myfuncNoConv);; time;
+GASMAN("collect"); ListX([1..10^5], [1..10], blafuncNoConv);; time;
```
Comment (fingolfin) [inline JuliaInterface/example/function_perform.gi]: Those `time;` statements are only valid in the interpreter; putting them into `.g` or `.gi` file is not really appropriate.

File: JuliaInterface/example/function_perform.jl [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    return a;
+end
+
+function myfunc( a::Int64, b::Int64 )
+    return a;
+end
+
+end
+
+function blafunc(a, b)
```
Comment (fingolfin) [inline JuliaInterface/example/function_perform.jl]: I understand that this is just my hacked together example, but if this is added to the repo, perhaps let's choose some better names? E.g. 
* Stuff -> function_perform to match the file (and BTW, perhaps use `function_calls` instead? That'd at least match the GAP nomenclature, and also fits with the `jl_call*` function names
* myFunc -> returnFirstWithMethodDispatch
* blafunc -> returnFirstGeneric

Well, or something along these lines.

File: JuliaInterface/example/function_perform.gi (hunk 9 above)
Comment (sebasguts) [reply JuliaInterface/example/function_perform.gi]: I will leave them in for now. I would prefer to have two versions of this file
1. a testfile
2. a version that I can copy&paste for demo proposes

**Comment by sebasguts** [comment]
I think I addressed all comments (except the clang-format one, we will do this in a single commit later).

File: JuliaInterface/configure.ac [hunk 11]
```diff
@@ -55,7 +55,7 @@ AC_ARG_WITH([julia],
 )
 JULIA_CPPFLAGS="-I$JULIA/include/julia"
 JULIA_LDPATH="$JULIA/lib"
-JULIA_LDFLAGS="-L$JULIA_LDPATH"
+JULIA_LDFLAGS="-L$JULIA_LDPATH -Wl,-rpath=$JULIA_LDPATH"
```
Comment (fingolfin) [inline JuliaInterface/configure.ac]: This is not needed anymore

**Pull request merged**